alg/los.cpp

```diff
@@ -0,0 +1,224 @@
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+// Test instructions:
+// * cmake -S . -B build -DBUILD_TESTING=ON -DCMAKE_BUILD_TYPE=Debug
+// * cmake --build build --target gdal_unit_test -j `nproc`
+// * ./build/autotest/cpp/gdal_unit_test --gtest_filter="test_alg.GDALIsLineOfSightVisible*"
+// Or, with GDB
+// *  gdb ./build/autotest/cpp/gdal_unit_test
+// * >>> r --gtest_filter="test_alg.GDALIsLineOfSightVisible*"
+
+#include <functional>
+#include <cmath>
+
+
+#include "cpl_port.h"
+#include "gdal_alg.h"
+
+// There's a plethora of bresenham implementations, all questionable production quality.
+// Bresenham optimizes for integer math, which makes sense for raster datasets in 2D.
+// For 3D, a 3D bresenham could be used if the altitude is also integer resolution.
+// 2D:
+// https://codereview.stackexchange.com/questions/77460/bresenhams-line-algorithm-optimization
+// https://gist.github.com/ssavi-ict/092501c69e2ffec65e96a8865470ad2f
+// https://blog.demofox.org/2015/01/17/bresenhams-drawing-algorithms/
+// https://en.wikipedia.org/wiki/Bresenham%27s_line_algorithm
+// https://www.cs.helsinki.fi/group/goa/mallinnus/lines/bresenh.html
+// https://stackoverflow.com/questions/10060046/drawing-lines-with-bresenhams-line-algorithm
+// http://www.edepot.com/linebresenham.html
+// 3D:
+// https://gist.github.com/yamamushi/5823518
+
+// Run bresenham terrain checking from (x1, y1) to (x2, y2).
+// The callback is run at every point along the line,
+// which should return True if the point is above terrain.
+// Bresenham2D will return true if all points have LOS between the start and end.
+bool Bresenham2D(int x1, int y1, int x2, int y2, std::function< auto(int, int) -> bool> OnBresenhamPoint) {
+    bool isAboveTerrain = true;
+	int	dx, dy;
+	int	incx, incy;
+
+	if (x2 >= x1)
+	{
+		dx = x2 - x1;
+		incx = 1;
+	}
+	else
+	{
+		dx = x1 - x2;
+		incx = -1;
+	}
+
+	if (y2 >= y1)
+	{
+		dy = y2 - y1;
+		incy = 1;
+	}
+	else
+	{
+		dy = y1 - y2;
+		incy = -1;
+	}
+
+	auto x = x1;
+	auto y = y1;
+    int	balance;
+
+	if (dx >= dy)
+	{
+		dy <<= 1;
```

Comment:
clearer and as efficient (the optimizer will generate the left shift for us)
```suggestion
		dy *= 2;
```

Reply:
nice. The left shift seems like a hold-over from much older compilers that couldn't optimize it.